libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,509 @@
+```
+
+We can't add the key to region 1 as it is full. Instead we fall back to trying
+region 2. If region 2 was full we would then try region 0 and so on.
+
+Adding key ONE would look like this:
+
+
+```
+0x000                  0x400                  0x800                 0xC00
+--------------------------------------------------------------------------
+|||||     Region 0     |||||     Region 1     |||||     Region 2     |||||
+|||||                  |||||                  |||||                  |||||
+|||||                  |||||KEY|EXAMPLE|TEST| |||||ONE               |||||
+--------------------------------------------------------------------------
+```
+
+### Finding keys from full regions
+
+Image the flash layout from above and then we wanted to find the non-existent
+key TWO.
+
+The hash of TWO indicates it should be in region 1. First we check region 1,
+but we don't find TWO. Next we try to find TWO in region 2. We don't kind the
+TWO object there, but as the region isn't empty we can't determine if it didn't
+fit. Next we try region 0. As this region is empty we stop looking.
+
+### Invalidating a key
+
+When the key ONE is invalidated with `invalidate_key()`, the only change in
+flash will be the `valid` flag. The object header for ONE will now look like:
+
+```
+0x400                                                                                              0x52C
+--------------------------------------------------------------------------------------------------------
+||||| version|len/flag|   len  |                              hashed_key                               |
+|||||        |        |        |        |        |        |        |        |        |        |        |
+|||||    0x00|00000000|    0x34|    0xed|    0xa1|    0x00|    0x78|    0x88|    0x61|    0x93|    0xbb|
+--------------------------------------------------------------------------------------------------------
+              ^
+```
+
+No changes will happen in flash until key TWO has also been invalidated.
+At which point `garbage_collect()` can erase the region.
+
+## Limitations of TickFS
+
+### Fragmentation
+
+Although TickFS has a `garbage_collect()` function, it makes no effort to
+handle fragmentation.
+
+That means that if you have the following objects in a region
+
+```
+0x000                  0x400
+----------------------------
+|||||     Region 0     |||||
+|||||                  |||||
+|||||ONE|TWO|THREE|FOUR|||||
+----------------------------
+```
+
+Where ONE, TWO and FOUR have been marked as invalid the entire region will still
+be classified as full. It will only not be full when all four objects are marked
+for invalid.
+
+It would be possible to have a reserved region and then free up the region with
+the following steps:
+ 1. Move the region to the reserved region
+ 1. Erase the above region
+ 1. Move the valid objects from the reserved region to the original region,
+    after being defraged in memory
+ 1. Erase the reserved region
```

Comment:
Does this mean that the reserved region is erased for every erase? I.e. it will fail first.

Reply:
This is just an example. I would expect that a reserved region would end up being used for every erase yes. It would be possible to move the reserved region (it changes every time to the last region you wanted to erase) but that then makes lookups slower as we loose ordering.